### DISCUSSION TITLE
How do I go about using exec to copy a file to a pod?

### DISCUSSION THREAD

**Discussion by cody-malcolm** [kubernetes-client/csharp#946] [Q&A] [answered]
I understand that copying files is [not a utility provided by the C# library](https://github.com/kubernetes-client/csharp/issues/464), but I was hoping to implement just the "copy from local to pod" functionality "manually" in my application. I more or less understand the Java implementation linked in the above discussion, however I haven't been able to get the streams right in my C# application to make it work. Are there any recommendations for how to approach this? If it's similar complexity, I'd prefer a blocking solution not an async one.

**Comment by tg123** [comment]
hmm you can share your code and let me see if I could help in the weekend

**Reply by tg123** [reply]
link #949

**Reply by cody-malcolm** [reply]
I'll keep an eye on #949. I made some progress and my current attempt uses the code below, I can successfully transfer a small text file however if I try a small binary file (4kb) it doesn't "arrive" and large binary files (~50mb) produce one of the following two errors:
"System.Net.Sockets.SocketException : An established connection was aborted by the software in your host machine"
"System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host"

My current code is as below. It is heavily based on the Java implementation, which makes no mention of any limitations with respect to file type or size.
```
using ICSharpCode.SharpZipLib.Tar;

//

string file = "path\\to\\my\\kubeconfig";

KubernetesClientConfiguration config = KubernetesClientConfiguration.BuildConfigFromConfigFile(file);
Kubernetes client = new Kubernetes(config);

string ns = "my-namespace";

k8s.Models.V1Pod pod = client.CoreV1.ListNamespacedPod(ns).Items.First();
string[] command = new string[] { "sh", "-c", "tar -xmf - -C " + "/app" }; // for now just hardcoding location

System.Net.WebSockets.WebSocket webSocket = await client.WebSocketNamespacedPodExecAsync(pod.Metadata.Name, ns, command,
    pod.Spec.Containers[0].Name).ConfigureAwait(false);

StreamDemuxer demux = new StreamDemuxer(webSocket);
demux.Start();

TarOutputStream archiveOutputStream = new TarOutputStream(demux.GetStream(null, ChannelIndex.StdIn), null);

string inputFileName = "path\\to\\realsmallfile.bin";

FileStream input = new FileStream(inputFileName, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 4096, useAsync: true);

TarEntry tarEntry = TarEntry.CreateEntryFromFile(inputFileName);
tarEntry.Name = "test.bin";

archiveOutputStream.PutNextEntry(tarEntry);
await input.CopyToAsync(archiveOutputStream).ConfigureAwait(false);
archiveOutputStream.CloseEntry();

archiveOutputStream.Close();
input.Close();
```
I originally tried it with the Stream.CopyTo(...) method, that has the same results.

**Comment by tg123** [comment, marked as answer]
thanks to @lazarkrstic 
#962 we had a cp example

**Reply by cody-malcolm** [reply]
Unfortunately some of the methods used are not available in KubernetesClient.Classic (such as ExecAsyncCallback, WriteAsync, FlushAsync). What I would like to know is if the cp example has been tested with binary files and/or large files? Also, can the classes that implement the above methods be included in the Classic library?

Thanks.

**Reply by tg123** [reply]
i will verify it, fix the websock chunk bug and create a kubectl version
the example should work with classic, but i will double confirm later
